ICE: trimmed_def_paths constructed, assertion failed: !new_self_ty.has_escaping_bound_vars()

### Code

Unfortunately I'm unable to reduce the testcase, but you can run it by opening https://github.com/Manishearth/icu4x/tree/ice-trimmed-def-paths. and from the `components/properties` directory, running `cargo test --doc ascii_hex_digit`. This code is in the midst of a refactor so most of the other tests will not pass.

I _suspect_ it has to do with the complicated trait stuff going on in provider.rs interacting with the docs.rs/yoke crate (cc @jackh726 @eddyb) but I'm not sure.




### Meta

Bug occurs on 1.60 stable and current nightly.

`rustc --version --verbose`:
```
$ rustc +nightly --version --verbose
rustc 1.62.0-nightly (311e2683e 2022-04-18)
binary: rustc
commit-hash: 311e2683e1bad87715b1558f7900e294d24ce491
commit-date: 2022-04-18
host: x86_64-unknown-linux-gnu
release: 1.62.0-nightly
LLVM version: 14.0.1

```

### Error output

```
thread 'rustc' panicked at 'assertion failed: !new_self_ty.has_escaping_bound_vars()', compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:1959:9
...
#0 [typeck] type-checking `main::_doctest_main_src_sets_rs_67_0`
#1 [typeck_item_bodies] type-checking all item bodies
#2 [analysis] running analysis passes on this crate
error: internal compiler error: no warnings or errors encountered even though `delayed_good_path_bugs` issued

error: internal compiler error: trimmed_def_paths constructed

```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
---- src/sets.rs - sets::get_ascii_hex_digit (line 67) stdout ----
thread 'rustc' panicked at 'assertion failed: !new_self_ty.has_escaping_bound_vars()', compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:1959:9
stack backtrace:
   0:     0x7f849eae182d - std::backtrace_rs::backtrace::libunwind::trace::h9d7060c592fe8069
                               at /rustc/311e2683e1bad87715b1558f7900e294d24ce491/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7f849eae182d - std::backtrace_rs::backtrace::trace_unsynchronized::hde51808b1556e5d1
                               at /rustc/311e2683e1bad87715b1558f7900e294d24ce491/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f849eae182d - std::sys_common::backtrace::_print_fmt::hd954a2d1cd9fdbd5
                               at /rustc/311e2683e1bad87715b1558f7900e294d24ce491/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x7f849eae182d - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h6271b22d4e4423c8
                               at /rustc/311e2683e1bad87715b1558f7900e294d24ce491/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x7f849eb3cecc - core::fmt::write::heb1c2a4e59b134a9
                               at /rustc/311e2683e1bad87715b1558f7900e294d24ce491/library/core/src/fmt/mod.rs:1194:17
   5:     0x7f849ead3011 - std::io::Write::write_fmt::h6652729f1feeae87
                               at /rustc/311e2683e1bad87715b1558f7900e294d24ce491/library/std/src/io/mod.rs:1655:15
   6:     0x7f849eae4545 - std::sys_common::backtrace::_print::ha9ae153024491f3f
                               at /rustc/311e2683e1bad87715b1558f7900e294d24ce491/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x7f849eae4545 - std::sys_common::backtrace::print::h110c021885084d6a
                               at /rustc/311e2683e1bad87715b1558f7900e294d24ce491/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x7f849eae4545 - std::panicking::default_hook::{{closure}}::h68d23f0b88a08886
                               at /rustc/311e2683e1bad87715b1558f7900e294d24ce491/library/std/src/panicking.rs:295:22
   9:     0x7f849eae41b9 - std::panicking::default_hook::h2e20223d1536191d
                               at /rustc/311e2683e1bad87715b1558f7900e294d24ce491/library/std/src/panicking.rs:314:9
  10:     0x7f849f286731 - rustc_driver[46ea51695bd673e5]::DEFAULT_HOOK::{closure#0}::{closure#0}
  11:     0x7f849eae4d16 - std::panicking::rust_panic_with_hook::hfe1e95a000a88a64
                               at /rustc/311e2683e1bad87715b1558f7900e294d24ce491/library/std/src/panicking.rs:702:17
  12:     0x7f849eae4ad9 - std::panicking::begin_panic_handler::{{closure}}::h04cf82910c0a653e
                               at /rustc/311e2683e1bad87715b1558f7900e294d24ce491/library/std/src/panicking.rs:586:13
  13:     0x7f849eae1ce4 - std::sys_common::backtrace::__rust_end_short_backtrace::h187f35deeb28abc2
                               at /rustc/311e2683e1bad87715b1558f7900e294d24ce491/library/std/src/sys_common/backtrace.rs:138:18
  14:     0x7f849eae4849 - rust_begin_unwind
                               at /rustc/311e2683e1bad87715b1558f7900e294d24ce491/library/std/src/panicking.rs:584:5
  15:     0x7f849eaa9ba3 - core::panicking::panic_fmt::h70547f234f37fa30
                               at /rustc/311e2683e1bad87715b1558f7900e294d24ce491/library/core/src/panicking.rs:142:14
  16:     0x7f849eaa9a6d - core::panicking::panic::hb8b030a543bc41a9
                               at /rustc/311e2683e1bad87715b1558f7900e294d24ce491/library/core/src/panicking.rs:48:5
  17:     0x7f84a00ea5d5 - <rustc_infer[a51b2704175fce84]::infer::InferCtxt as rustc_trait_selection[69015c12cf27210f]::traits::error_reporting::InferCtxtPrivExt>::mk_trait_obligation_with_new_self_ty
  18:     0x7f84a00ce857 - <rustc_infer[a51b2704175fce84]::infer::InferCtxt as rustc_trait_selection[69015c12cf27210f]::traits::error_reporting::suggestions::InferCtxtExt>::suggest_add_reference_to_arg::{closure#1}
  19:     0x7f84a00ce605 - <rustc_infer[a51b2704175fce84]::infer::InferCtxt as rustc_trait_selection[69015c12cf27210f]::traits::error_reporting::suggestions::InferCtxtExt>::suggest_add_reference_to_arg
  20:     0x7f84a00e4200 - <rustc_infer[a51b2704175fce84]::infer::InferCtxt as rustc_trait_selection[69015c12cf27210f]::traits::error_reporting::InferCtxtExt>::report_selection_error
  21:     0x7f84a00ec751 - <rustc_infer[a51b2704175fce84]::infer::InferCtxt as rustc_trait_selection[69015c12cf27210f]::traits::error_reporting::InferCtxtPrivExt>::report_fulfillment_error
  22:     0x7f84a00e0ac7 - <rustc_infer[a51b2704175fce84]::infer::InferCtxt as rustc_trait_selection[69015c12cf27210f]::traits::error_reporting::InferCtxtExt>::report_fulfillment_errors
  23:     0x7f84a0712792 - <rustc_typeck[1bd8200d7af772bc]::check::fn_ctxt::FnCtxt>::check_argument_types
  24:     0x7f84a06f7f6e - <rustc_typeck[1bd8200d7af772bc]::check::fn_ctxt::FnCtxt>::confirm_builtin_call
  25:     0x7f84a06f4438 - <rustc_typeck[1bd8200d7af772bc]::check::fn_ctxt::FnCtxt>::check_call
  26:     0x7f84a074a18d - <rustc_typeck[1bd8200d7af772bc]::check::fn_ctxt::FnCtxt>::check_expr_kind
  27:     0x7f84a0748fc0 - <rustc_typeck[1bd8200d7af772bc]::check::fn_ctxt::FnCtxt>::check_expr_with_expectation
  28:     0x7f84a074a305 - <rustc_typeck[1bd8200d7af772bc]::check::fn_ctxt::FnCtxt>::check_expr_kind
  29:     0x7f84a0748fc0 - <rustc_typeck[1bd8200d7af772bc]::check::fn_ctxt::FnCtxt>::check_expr_with_expectation
  30:     0x7f84a0737439 - <rustc_typeck[1bd8200d7af772bc]::check::fn_ctxt::FnCtxt>::check_stmt
  31:     0x7f84a0738c42 - <rustc_typeck[1bd8200d7af772bc]::check::fn_ctxt::FnCtxt>::check_block_with_expected
  32:     0x7f84a074a2b3 - <rustc_typeck[1bd8200d7af772bc]::check::fn_ctxt::FnCtxt>::check_expr_kind
  33:     0x7f84a0748fc0 - <rustc_typeck[1bd8200d7af772bc]::check::fn_ctxt::FnCtxt>::check_expr_with_expectation
  34:     0x7f84a08622d6 - rustc_typeck[1bd8200d7af772bc]::check::check::check_fn
  35:     0x7f84a07c2140 - <rustc_infer[a51b2704175fce84]::infer::InferCtxtBuilder>::enter::<&rustc_middle[4038938c89644caa]::ty::context::TypeckResults, <rustc_typeck[1bd8200d7af772bc]::check::inherited::InheritedBuilder>::enter<rustc_typeck[1bd8200d7af772bc]::check::typeck_with_fallback<rustc_typeck[1bd8200d7af772bc]::check::typeck::{closure#0}>::{closure#1}, &rustc_middle[4038938c89644caa]::ty::context::TypeckResults>::{closure#0}>
  36:     0x7f84a0794f00 - rustc_typeck[1bd8200d7af772bc]::check::typeck
  37:     0x7f84a0ca9390 - rustc_query_system[103384a90f76de31]::query::plumbing::try_execute_query::<rustc_query_impl[8b18bc69f52059fc]::plumbing::QueryCtxt, rustc_query_system[103384a90f76de31]::query::caches::DefaultCache<rustc_span[8638436c65ab17fb]::def_id::LocalDefId, &rustc_middle[4038938c89644caa]::ty::context::TypeckResults>>
  38:     0x7f84a0dadb08 - <rustc_query_impl[8b18bc69f52059fc]::Queries as rustc_middle[4038938c89644caa]::ty::query::QueryEngine>::typeck
  39:     0x7f84a0846c18 - <rustc_middle[4038938c89644caa]::hir::map::Map>::par_body_owners::<rustc_typeck[1bd8200d7af772bc]::check::typeck_item_bodies::{closure#0}>
  40:     0x7f84a156123c - rustc_typeck[1bd8200d7af772bc]::check::typeck_item_bodies
  41:     0x7f84a17b5c92 - rustc_query_system[103384a90f76de31]::query::plumbing::try_execute_query::<rustc_query_impl[8b18bc69f52059fc]::plumbing::QueryCtxt, rustc_query_system[103384a90f76de31]::query::caches::DefaultCache<(), ()>>
  42:     0x7f84a17dd4a1 - rustc_query_system[103384a90f76de31]::query::plumbing::get_query::<rustc_query_impl[8b18bc69f52059fc]::queries::typeck_item_bodies, rustc_query_impl[8b18bc69f52059fc]::plumbing::QueryCtxt>
  43:     0x7f84a15702f3 - <rustc_session[5849b3ac9fc37c68]::session::Session>::time::<(), rustc_typeck[1bd8200d7af772bc]::check_crate::{closure#7}>
  44:     0x7f84a156de93 - rustc_typeck[1bd8200d7af772bc]::check_crate
  45:     0x7f84a12ee477 - rustc_interface[668261b8f47f72c0]::passes::analysis
  46:     0x7f84a17abdc4 - rustc_query_system[103384a90f76de31]::query::plumbing::try_execute_query::<rustc_query_impl[8b18bc69f52059fc]::plumbing::QueryCtxt, rustc_query_system[103384a90f76de31]::query::caches::DefaultCache<(), core[2fe9dabce7a4eee7]::result::Result<(), rustc_errors[87e4665ca07ec85e]::ErrorGuaranteed>>>
  47:     0x7f84a17ed24e - rustc_query_system[103384a90f76de31]::query::plumbing::get_query::<rustc_query_impl[8b18bc69f52059fc]::queries::analysis, rustc_query_impl[8b18bc69f52059fc]::plumbing::QueryCtxt>
  48:     0x7f84a12cbaf7 - <rustc_interface[668261b8f47f72c0]::passes::QueryContext>::enter::<rustc_driver[46ea51695bd673e5]::run_compiler::{closure#1}::{closure#2}::{closure#3}, core[2fe9dabce7a4eee7]::result::Result<(), rustc_errors[87e4665ca07ec85e]::ErrorGuaranteed>>
  49:     0x7f84a12b7078 - <rustc_interface[668261b8f47f72c0]::interface::Compiler>::enter::<rustc_driver[46ea51695bd673e5]::run_compiler::{closure#1}::{closure#2}, core[2fe9dabce7a4eee7]::result::Result<core[2fe9dabce7a4eee7]::option::Option<rustc_interface[668261b8f47f72c0]::queries::Linker>, rustc_errors[87e4665ca07ec85e]::ErrorGuaranteed>>
  50:     0x7f84a12dfe3f - rustc_span[8638436c65ab17fb]::with_source_map::<core[2fe9dabce7a4eee7]::result::Result<(), rustc_errors[87e4665ca07ec85e]::ErrorGuaranteed>, rustc_interface[668261b8f47f72c0]::interface::create_compiler_and_run<core[2fe9dabce7a4eee7]::result::Result<(), rustc_errors[87e4665ca07ec85e]::ErrorGuaranteed>, rustc_driver[46ea51695bd673e5]::run_compiler::{closure#1}>::{closure#1}>
  51:     0x7f84a12b7c97 - <scoped_tls[b360d135f6f091a9]::ScopedKey<rustc_span[8638436c65ab17fb]::SessionGlobals>>::set::<rustc_interface[668261b8f47f72c0]::interface::run_compiler<core[2fe9dabce7a4eee7]::result::Result<(), rustc_errors[87e4665ca07ec85e]::ErrorGuaranteed>, rustc_driver[46ea51695bd673e5]::run_compiler::{closure#1}>::{closure#0}, core[2fe9dabce7a4eee7]::result::Result<(), rustc_errors[87e4665ca07ec85e]::ErrorGuaranteed>>
  52:     0x7f84a12cd1df - std[4fdfe6e28689bcfc]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[668261b8f47f72c0]::util::run_in_thread_pool_with_globals<rustc_interface[668261b8f47f72c0]::interface::run_compiler<core[2fe9dabce7a4eee7]::result::Result<(), rustc_errors[87e4665ca07ec85e]::ErrorGuaranteed>, rustc_driver[46ea51695bd673e5]::run_compiler::{closure#1}>::{closure#0}, core[2fe9dabce7a4eee7]::result::Result<(), rustc_errors[87e4665ca07ec85e]::ErrorGuaranteed>>::{closure#0}, core[2fe9dabce7a4eee7]::result::Result<(), rustc_errors[87e4665ca07ec85e]::ErrorGuaranteed>>
  53:     0x7f84a12cd329 - <<std[4fdfe6e28689bcfc]::thread::Builder>::spawn_unchecked_<rustc_interface[668261b8f47f72c0]::util::run_in_thread_pool_with_globals<rustc_interface[668261b8f47f72c0]::interface::run_compiler<core[2fe9dabce7a4eee7]::result::Result<(), rustc_errors[87e4665ca07ec85e]::ErrorGuaranteed>, rustc_driver[46ea51695bd673e5]::run_compiler::{closure#1}>::{closure#0}, core[2fe9dabce7a4eee7]::result::Result<(), rustc_errors[87e4665ca07ec85e]::ErrorGuaranteed>>::{closure#0}, core[2fe9dabce7a4eee7]::result::Result<(), rustc_errors[87e4665ca07ec85e]::ErrorGuaranteed>>::{closure#1} as core[2fe9dabce7a4eee7]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  54:     0x7f849eaeebc3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h6caff75e9500dfec
                               at /rustc/311e2683e1bad87715b1558f7900e294d24ce491/library/alloc/src/boxed.rs:1866:9
  55:     0x7f849eaeebc3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hfcb7dd749eba4870
                               at /rustc/311e2683e1bad87715b1558f7900e294d24ce491/library/alloc/src/boxed.rs:1866:9
  56:     0x7f849eaeebc3 - std::sys::unix::thread::Thread::new::thread_start::h865210c19e650802
                               at /rustc/311e2683e1bad87715b1558f7900e294d24ce491/library/std/src/sys/unix/thread.rs:108:17
  57:     0x7f849ea0fd80 - start_thread
  58:     0x7f849e91e76f - clone
  59:                0x0 - <unknown>

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.62.0-nightly (311e2683e 2022-04-18) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type bin -C codegen-units=1 -C embed-bitcode=no

query stack during panic:
#0 [typeck] type-checking `main::_doctest_main_src_sets_rs_67_0`
#1 [typeck_item_bodies] type-checking all item bodies
#2 [analysis] running analysis passes on this crate
end of query stack

error: internal compiler error: no warnings or errors encountered even though `delayed_good_path_bugs` issued

error: internal compiler error: trimmed_def_paths constructed

```

</p>
</details>



<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"ricked-twice"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

As a quick test hack: if one bails out of the suggestion when there are escaping bound vars, this code would emit "error[E0277]: the trait bound `for<'de> CodePointSet<'de>: Deserialize<'_>` is not satisfied".

--- a/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs
+++ b/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs
@@ -1384,8 +1384,7 @@ trait InferCtxtPrivExt<'hir, 'tcx> {
     fn mk_trait_obligation_with_new_self_ty(
         &self,
         param_env: ty::ParamEnv<'tcx>,
-        trait_ref: ty::PolyTraitPredicate<'tcx>,
-        new_self_ty: Ty<'tcx>,
+        trait_ref_and_ty: ty::Binder<'tcx, (ty::TraitPredicate<'tcx>, Ty<'tcx>)>,
     ) -> PredicateObligation<'tcx>;
 
     fn maybe_report_ambiguity(
@@ -1923,14 +1922,11 @@ impl<'a, 'tcx> InferCtxtPrivExt<'a, 'tcx> for InferCtxt<'a, 'tcx> {
     fn mk_trait_obligation_with_new_self_ty(
         &self,
         param_env: ty::ParamEnv<'tcx>,
-        trait_ref: ty::PolyTraitPredicate<'tcx>,
-        new_self_ty: Ty<'tcx>,
+        trait_ref_and_ty: ty::Binder<'tcx, (ty::TraitPredicate<'tcx>, Ty<'tcx>)>,
     ) -> PredicateObligation<'tcx> {
-        assert!(!new_self_ty.has_escaping_bound_vars());
-
-        let trait_pred = trait_ref.map_bound_ref(|tr| ty::TraitPredicate {
+        let trait_pred = trait_ref_and_ty.map_bound_ref(|(tr, new_self_ty)| ty::TraitPredicate {
             trait_ref: ty::TraitRef {
-                substs: self.tcx.mk_substs_trait(new_self_ty, &tr.trait_ref.substs[1..]),
+                substs: self.tcx.mk_substs_trait(*new_self_ty, &tr.trait_ref.substs[1..]),
                 ..tr.trait_ref
             },
             ..*tr

--- a/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
+++ b/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
@@ -637,8 +637,10 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                 if let Some(steps) = autoderef.find_map(|(ty, steps)| {
                     // Re-add the `&`
                     let ty = self.tcx.mk_ref(region, TypeAndMut { ty, mutbl });
-                    let obligation =
-                        self.mk_trait_obligation_with_new_self_ty(param_env, real_trait_pred, ty);
+                    let real_trait_pred_and_ty =
+                        real_trait_pred.map_bound(|inner_trait_pred| (inner_trait_pred, ty));
+                    let obligation = self
+                        .mk_trait_obligation_with_new_self_ty(param_env, real_trait_pred_and_ty);
                     Some(steps).filter(|_| self.predicate_may_hold(&obligation))
                 }) {
                     if steps > 0 {
@@ -659,10 +661,11 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                     }
                 } else if real_trait_pred != trait_pred {
                     // This branch addresses #87437.
+                    let real_trait_pred_and_base_ty =
+                        real_trait_pred.map_bound(|inner_trait_pred| (inner_trait_pred, base_ty));
                     let obligation = self.mk_trait_obligation_with_new_self_ty(
                         param_env,
-                        real_trait_pred,
-                        base_ty,
+                        real_trait_pred_and_base_ty,
                     );
                     if self.predicate_may_hold(&obligation) {
                         err.span_suggestion_verbose(
@@ -720,9 +723,7 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
         err: &mut Diagnostic,
         trait_pred: ty::PolyTraitPredicate<'tcx>,
     ) -> bool {
-        let Some(self_ty) = trait_pred.self_ty().no_bound_vars() else {
-            return false;
-        };
+        let self_ty = trait_pred.self_ty().skip_binder();
 
         let (def_id, output_ty, callable) = match *self_ty.kind() {
             ty::Closure(def_id, substs) => (def_id, substs.as_closure().sig().output(), "closure"),
@@ -731,14 +732,12 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
         };
         let msg = format!("use parentheses to call the {}", callable);
 
-        // `mk_trait_obligation_with_new_self_ty` only works for types with no escaping bound
-        // variables, so bail out if we have any.
-        let Some(output_ty) = output_ty.no_bound_vars() else {
-            return false;
-        };
+        let output_ty = self.tcx.liberate_late_bound_regions(def_id, output_ty);
+
+        let trait_pred_and_self = trait_pred.map_bound(|trait_pred| (trait_pred, output_ty));
 
         let new_obligation =
-            self.mk_trait_obligation_with_new_self_ty(obligation.param_env, trait_pred, output_ty);
+            self.mk_trait_obligation_with_new_self_ty(obligation.param_env, trait_pred_and_self);
 
         match self.evaluate_obligation(&new_obligation) {
             Ok(
@@ -842,96 +841,102 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
         let param_env = obligation.param_env;
 
         // Try to apply the original trait binding obligation by borrowing.
-        let mut try_borrowing = |old_pred: ty::PolyTraitPredicate<'tcx>,
-                                 blacklist: &[DefId]|
-         -> bool {
-            if blacklist.contains(&old_pred.def_id()) {
-                return false;
-            }
-
-            // This is a quick fix to resolve an ICE (#96223).
-            // This change should probably be deeper.
-            // As suggested by @jackh726, `mk_trait_obligation_with_new_self_ty` could take a `Binder<(TraitRef, Ty)>
-            // instead of `Binder<Ty>` leading to some changes to its call places.
-            let Some(orig_ty) = old_pred.self_ty().no_bound_vars() else {
-                return false;
-            };
-            let mk_result = |new_ty| {
-                let obligation =
-                    self.mk_trait_obligation_with_new_self_ty(param_env, old_pred, new_ty);
-                self.predicate_must_hold_modulo_regions(&obligation)
-            };
-            let imm_result = mk_result(self.tcx.mk_imm_ref(self.tcx.lifetimes.re_static, orig_ty));
-            let mut_result = mk_result(self.tcx.mk_mut_ref(self.tcx.lifetimes.re_static, orig_ty));
-
-            if imm_result || mut_result {
-                if let Ok(snippet) = self.tcx.sess.source_map().span_to_snippet(span) {
-                    // We have a very specific type of error, where just borrowing this argument
-                    // might solve the problem. In cases like this, the important part is the
-                    // original type obligation, not the last one that failed, which is arbitrary.
-                    // Because of this, we modify the error to refer to the original obligation and
-                    // return early in the caller.
-
-                    let msg = format!(
-                        "the trait bound `{}: {}` is not satisfied",
-                        orig_ty,
-                        old_pred.print_modifiers_and_trait_path(),
-                    );
-                    if has_custom_message {
-                        err.note(&msg);
-                    } else {
-                        err.message =
-                            vec![(rustc_errors::DiagnosticMessage::Str(msg), Style::NoStyle)];
-                    }
-                    if snippet.starts_with('&') {
-                        // This is already a literal borrow and the obligation is failing
-                        // somewhere else in the obligation chain. Do not suggest non-sense.
-                        return false;
-                    }
-                    err.span_label(
-                        span,
-                        &format!(
-                            "expected an implementor of trait `{}`",
-                            old_pred.print_modifiers_and_trait_path(),
-                        ),
-                    );
-
-                    // This if is to prevent a special edge-case
-                    if matches!(
-                        span.ctxt().outer_expn_data().kind,
-                        ExpnKind::Root | ExpnKind::Desugaring(DesugaringKind::ForLoop)
-                    ) {
-                        // We don't want a borrowing suggestion on the fields in structs,
-                        // ```
-                        // struct Foo {
-                        //  the_foos: Vec<Foo>
-                        // }
-                        // ```
-
-                        if imm_result && mut_result {
-                            err.span_suggestions(
-                                span.shrink_to_lo(),
-                                "consider borrowing here",
-                                ["&".to_string(), "&mut ".to_string()].into_iter(),
-                                Applicability::MaybeIncorrect,
-                            );
-                        } else {
-                            err.span_suggestion_verbose(
-                                span.shrink_to_lo(),
-                                &format!(
-                                    "consider{} borrowing here",
-                                    if mut_result { " mutably" } else { "" }
-                                ),
-                                format!("&{}", if mut_result { "mut " } else { "" }),
-                                Applicability::MaybeIncorrect,
-                            );
-                        }
-                    }
-                    return true;
+        let mut try_borrowing =
+            |old_pred: ty::PolyTraitPredicate<'tcx>, blacklist: &[DefId]| -> bool {
+                if blacklist.contains(&old_pred.def_id()) {
+                    return false;
                 }
-            }
-            return false;
-        };
+                // We map bounds to `&T` and `&mut T`
+                let trait_pred_and_imm_ref = old_pred.map_bound(|trait_pred| {
+                    (
+                        trait_pred,
+                        self.tcx.mk_imm_ref(self.tcx.lifetimes.re_static, trait_pred.self_ty()),
+                    )
+                });
+                let trait_pred_and_mut_ref = old_pred.map_bound(|trait_pred| {
+                    (
+                        trait_pred,
+                        self.tcx.mk_mut_ref(self.tcx.lifetimes.re_static, trait_pred.self_ty()),
+                    )
+                });
+
+                let mk_result = |trait_pred_and_new_ty| {
+                    let obligation =
+                        self.mk_trait_obligation_with_new_self_ty(param_env, trait_pred_and_new_ty);
+                    self.predicate_must_hold_modulo_regions(&obligation)
+                };
+                let imm_result = mk_result(trait_pred_and_imm_ref);
+                let mut_result = mk_result(trait_pred_and_mut_ref);
+
+                if imm_result || mut_result {
+                    if let Ok(snippet) = self.tcx.sess.source_map().span_to_snippet(span) {
+                        // We have a very specific type of error, where just borrowing this argument
+                        // might solve the problem. In cases like this, the important part is the
+                        // original type obligation, not the last one that failed, which is arbitrary.
+                        // Because of this, we modify the error to refer to the original obligation and
+                        // return early in the caller.
+
+                        let msg = format!(
+                            "the trait bound `{}: {}` is not satisfied",
+                            // Safe to skip binder here
+                            old_pred.self_ty().skip_binder(),
+                            old_pred.print_modifiers_and_trait_path(),
+                        );
+                        if has_custom_message {
+                            err.note(&msg);
+                        } else {
+                            err.message =
+                                vec![(rustc_errors::DiagnosticMessage::Str(msg), Style::NoStyle)];
+                        }
+                        if snippet.starts_with('&') {
+                            // This is already a literal borrow and the obligation is failing
+                            // somewhere else in the obligation chain. Do not suggest non-sense.
+                            return false;
+                        }
+                        err.span_label(
+                            span,
+                            &format!(
+                                "expected an implementor of trait `{}`",
+                                old_pred.print_modifiers_and_trait_path(),
+                            ),
+                        );
+
+                        // This if is to prevent a special edge-case
+                        if matches!(
+                            span.ctxt().outer_expn_data().kind,
+                            ExpnKind::Root | ExpnKind::Desugaring(DesugaringKind::ForLoop)
+                        ) {
+                            // We don't want a borrowing suggestion on the fields in structs,
+                            // ```
+                            // struct Foo {
+                            //  the_foos: Vec<Foo>
+                            // }
+                            // ```
+
+                            if imm_result && mut_result {
+                                err.span_suggestions(
+                                    span.shrink_to_lo(),
+                                    "consider borrowing here",
+                                    ["&".to_string(), "&mut ".to_string()].into_iter(),
+                                    Applicability::MaybeIncorrect,
+                                );
+                            } else {
+                                err.span_suggestion_verbose(
+                                    span.shrink_to_lo(),
+                                    &format!(
+                                        "consider{} borrowing here",
+                                        if mut_result { " mutably" } else { "" }
+                                    ),
+                                    format!("&{}", if mut_result { "mut " } else { "" }),
+                                    Applicability::MaybeIncorrect,
+                                );
+                            }
+                        }
+                        return true;
+                    }
+                }
+                return false;
+            };
 
         if let ObligationCauseCode::ImplDerivedObligation(cause) = &*code {
             try_borrowing(cause.derived.parent_trait_pred, &[])
@@ -992,9 +997,8 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                 return false;
             }
 
-            let Some(mut suggested_ty) = trait_pred.self_ty().no_bound_vars() else {
-                return false;
-            };
+            // We skip binder here
+            let mut suggested_ty = trait_pred.self_ty().skip_binder();
 
             for refs_remaining in 0..refs_number {
                 let ty::Ref(_, inner_ty, _) = suggested_ty.kind() else {
@@ -1002,10 +1006,13 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                 };
                 suggested_ty = *inner_ty;
 
+                // We remap bounds here
+                let trait_pred_and_suggested_ty =
+                    trait_pred.map_bound(|trait_pred| (trait_pred, suggested_ty));
+
                 let new_obligation = self.mk_trait_obligation_with_new_self_ty(
                     obligation.param_env,
-                    trait_pred,
-                    suggested_ty,
+                    trait_pred_and_suggested_ty,
                 );
 
                 if self.predicate_may_hold(&new_obligation) {
@@ -1141,10 +1148,12 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                     hir::Mutability::Not => self.tcx.mk_mut_ref(region, t_type),
                 };
 
+                let trait_pred_and_suggested_ty =
+                    trait_pred.map_bound(|trait_pred| (trait_pred, suggested_ty));
+
                 let new_obligation = self.mk_trait_obligation_with_new_self_ty(
                     obligation.param_env,
-                    trait_pred,
-                    suggested_ty,
+                    trait_pred_and_suggested_ty,
                 );
                 let suggested_ty_would_satisfy_obligation = self
                     .evaluate_obligation_no_overflow(&new_obligation)
@@ -1195,7 +1204,9 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
             // Only suggest this if the expression behind the semicolon implements the predicate
             && let Some(typeck_results) = self.in_progress_typeck_results
             && let Some(ty) = typeck_results.borrow().expr_ty_opt(expr)
-            && self.predicate_may_hold(&self.mk_trait_obligation_with_new_self_ty(obligation.param_env, trait_pred, ty))
+            && self.predicate_may_hold(&self.mk_trait_obligation_with_new_self_ty(
+                obligation.param_env, trait_pred.map_bound(|trait_pred| (trait_pred, ty))
+            ))
         {
             err.span_label(
                 expr.span,
@@ -2727,8 +2738,7 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                 );
                 let try_obligation = self.mk_trait_obligation_with_new_self_ty(
                     obligation.param_env,
-                    trait_pred,
-                    normalized_ty.ty().unwrap(),
+                    trait_pred.map_bound(|trait_pred| (trait_pred, normalized_ty.ty().unwrap())),
                 );
                 debug!("suggest_await_before_try: try_trait_obligation {:?}", try_obligation);
                 if self.predicate_may_hold(&try_obligation)

--- a/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
+++ b/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
@@ -628,15 +628,17 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
             if let Some(parent_trait_pred) = parent_trait_pred {
                 real_trait_pred = parent_trait_pred;
             }
-            let Some(real_ty) = real_trait_pred.self_ty().no_bound_vars() else {
-                continue;
-            };
+
+            // Skipping binder here, remapping below
+            let real_ty = real_trait_pred.self_ty().skip_binder();
 
             if let ty::Ref(region, base_ty, mutbl) = *real_ty.kind() {
                 let mut autoderef = Autoderef::new(self, param_env, body_id, span, base_ty, span);
                 if let Some(steps) = autoderef.find_map(|(ty, steps)| {
                     // Re-add the `&`
                     let ty = self.tcx.mk_ref(region, TypeAndMut { ty, mutbl });
+
+                    // Remapping bound vars here
                     let real_trait_pred_and_ty =
                         real_trait_pred.map_bound(|inner_trait_pred| (inner_trait_pred, ty));
                     let obligation = self
@@ -661,6 +663,8 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                     }
                 } else if real_trait_pred != trait_pred {
                     // This branch addresses #87437.
+
+                    // Remapping bound vars here
                     let real_trait_pred_and_base_ty =
                         real_trait_pred.map_bound(|inner_trait_pred| (inner_trait_pred, base_ty));
                     let obligation = self.mk_trait_obligation_with_new_self_ty(
@@ -723,6 +727,7 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
         err: &mut Diagnostic,
         trait_pred: ty::PolyTraitPredicate<'tcx>,
     ) -> bool {
+        // Skipping binder here, remapping below
         let self_ty = trait_pred.self_ty().skip_binder();
 
         let (def_id, output_ty, callable) = match *self_ty.kind() {
@@ -732,8 +737,11 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
         };
         let msg = format!("use parentheses to call the {}", callable);
 
+        // "We should really create a single list of bound vars from the combined vars
+        // from the predicate and function, but instead we just liberate the function bound vars"
         let output_ty = self.tcx.liberate_late_bound_regions(def_id, output_ty);
 
+        // Remapping bound vars here
         let trait_pred_and_self = trait_pred.map_bound(|trait_pred| (trait_pred, output_ty));
 
         let new_obligation =
@@ -876,12 +884,7 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                         // Because of this, we modify the error to refer to the original obligation and
                         // return early in the caller.
 
-                        let msg = format!(
-                            "the trait bound `{}: {}` is not satisfied",
-                            // Safe to skip binder here
-                            old_pred.self_ty().skip_binder(),
-                            old_pred.print_modifiers_and_trait_path(),
-                        );
+                        let msg = format!("the trait bound `{}` is not satisfied", old_pred);
                         if has_custom_message {
                             err.note(&msg);
                         } else {
@@ -997,7 +1000,7 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                 return false;
             }
 
-            // We skip binder here
+            // Skipping binder here, remapping below
             let mut suggested_ty = trait_pred.self_ty().skip_binder();
 
             for refs_remaining in 0..refs_number {
@@ -1006,7 +1009,7 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                 };
                 suggested_ty = *inner_ty;
 
-                // We remap bounds here
+                // Remapping bound vars here
                 let trait_pred_and_suggested_ty =
                     trait_pred.map_bound(|trait_pred| (trait_pred, suggested_ty));
 
@@ -1132,22 +1135,15 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                 return;
             }
 
+            // Skipping binder here, remapping below
             if let ty::Ref(region, t_type, mutability) = *trait_pred.skip_binder().self_ty().kind()
             {
-                if region.is_late_bound() || t_type.has_escaping_bound_vars() {
-                    // Avoid debug assertion in `mk_obligation_for_def_id`.
-                    //
-                    // If the self type has escaping bound vars then it's not
-                    // going to be the type of an expression, so the suggestion
-                    // probably won't apply anyway.
-                    return;
-                }
-
                 let suggested_ty = match mutability {
                     hir::Mutability::Mut => self.tcx.mk_imm_ref(region, t_type),
                     hir::Mutability::Not => self.tcx.mk_mut_ref(region, t_type),
                 };
 
+                // Remapping bound vars here
                 let trait_pred_and_suggested_ty =
                     trait_pred.map_bound(|trait_pred| (trait_pred, suggested_ty));
 

--- a/src/test/ui/binop/issue-77910-1.stderr
+++ b/src/test/ui/binop/issue-77910-1.stderr
@@ -12,20 +12,14 @@ LL |     assert_eq!(foo, y);
 error[E0277]: `for<'r> fn(&'r i32) -> &'r i32 {foo}` doesn't implement `Debug`
   --> $DIR/issue-77910-1.rs:8:5
    |
+LL | fn foo(s: &i32) -> &i32 {
+   |    --- consider calling this function
+...
 LL |     assert_eq!(foo, y);
    |     ^^^^^^^^^^^^^^^^^^ `for<'r> fn(&'r i32) -> &'r i32 {foo}` cannot be formatted using `{:?}` because it doesn't implement `Debug`
    |
    = help: the trait `Debug` is not implemented for `for<'r> fn(&'r i32) -> &'r i32 {foo}`
-   = help: the following other types implement trait `Debug`:
-             extern "C" fn() -> Ret
-             extern "C" fn(A) -> Ret
-             extern "C" fn(A, ...) -> Ret
-             extern "C" fn(A, B) -> Ret
-             extern "C" fn(A, B, ...) -> Ret
-             extern "C" fn(A, B, C) -> Ret
-             extern "C" fn(A, B, C, ...) -> Ret
-             extern "C" fn(A, B, C, D) -> Ret
-           and 68 others
+   = help: use parentheses to call the function: `foo(s)`
    = note: this error originates in the macro `assert_eq` (in Nightly builds, run with -Z macro-backtrace for more info)
 
 error: aborting due to 2 previous errors

--- a/src/test/ui/suggestions/imm-ref-trait-object-literal-bound-regions.stderr
+++ b/src/test/ui/suggestions/imm-ref-trait-object-literal-bound-regions.stderr
@@ -5,6 +5,7 @@ LL |     foo::<S>(s);
    |     ^^^^^^^^ the trait `for<'b> Trait` is not implemented for `&'b S`
    |
    = help: the trait `Trait` is implemented for `&'a mut S`
+   = note: `for<'b> Trait` is implemented for `&'b mut S`, but not for `&'b S`
 note: required by a bound in `foo`
   --> $DIR/imm-ref-trait-object-literal-bound-regions.rs:11:20
    |

